library/std/src/sys/uefi/args.rs

```diff
@@ -0,0 +1,152 @@
+use r_efi::protocols::loaded_image;
+
+use crate::env::current_exe;
+use crate::ffi::OsString;
+use crate::fmt;
+use crate::iter::Iterator;
+use crate::mem::size_of;
+use crate::sys::uefi::helpers;
+use crate::vec;
+
+pub struct Args {
+    parsed_args_list: vec::IntoIter<OsString>,
+}
+
+pub fn args() -> Args {
+    // SAFETY: Each loaded image has an image handle that supports EFI_LOADED_IMAGE_PROTOCOL
+    let protocol =
+        helpers::image_handle_protocol::<loaded_image::Protocol>(loaded_image::PROTOCOL_GUID)
+            .unwrap();
+
+    let lp_size = unsafe { (*protocol.as_ptr()).load_options_size } as usize;
+
+    // Break if we are sure that it cannot be UTF-16
+    if lp_size < size_of::<u16>() || lp_size % size_of::<u16>() != 0 {
+        return Args {
+            parsed_args_list: Vec::from([current_exe().map(Into::into).unwrap_or_default()])
+                .into_iter(),
+        };
+    }
+
+    let lp_cmd_line = unsafe {
+        let lp_size = lp_size / size_of::<u16>();
+        let temp = (*protocol.as_ptr()).load_options as *const u16;
+        crate::slice::from_raw_parts(temp, lp_size)
+    };
+
+    let vec_args = parse_lp_cmd_line(lp_cmd_line)
+        .unwrap_or(Vec::from([current_exe().map(Into::into).unwrap_or_default()]));
```

Comment:
Is the expectation that the 0th argument passed in the protocol will be the current_exe() already, so we shouldn't prepend it? I think it would probably be a bit cleaner regardless to pull out the `Vec::from(...)` into a variable at the top of this function.

I'd also like some clarification on the intent of the unwrap_or_default(), which seems a bit weird. If we can't get current_exe(), should we be exposing that as an empty OsString? I would have naively expected just omitting it entirely, personally.

Reply:
Well, the windows implementation seemed to return `current_exe()` if parsing failed, so I was trying to do the same:
```rust
        let parsed_args_list = parse_lp_cmd_line(WStrUnits::new(lp_cmd_line), || {
            current_exe().map(PathBuf::into_os_string).unwrap_or_else(|_| OsString::new())
        });

```

Reply:
All of the unwrap_or's currently unconditionally run the Vec::from anyway, so it's not optimized anyway. If you wanted to save on the allocation you can do something like this:

```
let exe = || Vec::from([current_exe().map(Into::into).unwrap_or_default()]);
// ...

parse_lp_cmd_line(lp_cmd_line).unwrap_or_else(exe)
```

Reply:
If we're matching Windows here that seems ok.